### PR TITLE
DEV-3176 | Add support for new feature flag for mail chimp integration

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

This PR adds a new DRF permission, `HasFlaggedAccessToMailChimp` that can be used to gate access to arbitrary API endpoints based on whether or not user has this flag. 

This same flag will be available on the front end to gate access to componentry there, but this PR does not itself touch the front end.

This PR adds a temporary view that allows us to fully test the new permission through the view layer. In future PRs, we'll start using this permission to gate access to substantive resources.

#### Why are we doing this? How does it help us?

This is prelliminary work that will allow us to gate server- and SPA-side resources based on whether or not a user has the `"mailchimp-integration-access"` flag.

#### How should this be manually tested? Please include detailed step-by-step instructions.

Open an isolated incognito browser tab and go to [this page for the Mail chimp integration feature flag in the review app](https://dev-3176.revengine-review.org/nrhadmin/waffle/flag/4/change/) for this ticket. 

Ensure the flag has the following settings:

- The "Unknown" option should be chosen for "Everyone" 
- Superusers should be unchecked
- Staff should be unchecked
- There should be no individually assigned flag grantees.

Now in another isolated browser environment (such that cookies and auth tokens from previous steps don't leak), you'll go through the following sequence, visit the following URL: https://dev-3176.revengine-review.org/api/v1/mailchimp/ . You should get a 401 error.

Now back in the second browser environment, visit https://dev-3176.revengine-review.org/api/v1/mailchimp/ again. This time you'll also get a 401, as the endpoint has been set up to require authentication.

Next, in this same browser environment, sign in to the SPA using an account with superuser status. Back in the Django admin, turn the flag on for superusers. When you go to https://dev-3176.revengine-review.org/api/v1/mailchimp/ you should get a 200.

You can try with different combinations of users + flag settings (specifically, try out settings for everyone vs. superusers vs. staff user vs individually assigned users), using the same flow as above.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

No

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-3176](https://news-revenue-hub.atlassian.net/browse/DEV-3176)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

Across environments, after this PR is merged, we'll need to create a feature flag with the name "mailchimp-integration-access" in the Django admin. Initially, it can be created with the false for super users and staff, and "everyone" set to "Unknown".  

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No

[DEV-3176]: https://news-revenue-hub.atlassian.net/browse/DEV-3176?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/api/permissions.py [hunk 1]
```diff
@@ -113,7 +116,31 @@ def is_a_contributor(user):
...
+class HasFlaggedAccessToMailChimp(BaseFlaggedResourceAccess):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        Flag = get_waffle_flag_model()
+        self.flag = Flag.objects.filter(name=MAILCHIMP_INTEGRATION_ACCESS_FLAG_NAME).first()
+        if not self.flag:
+            raise ApiConfigurationError()
+
+    def __str__(self):
+        return f"`HasFlaggedAccessToMailChimp` via {self.flag.name}"
```
Comment (njharman) [inline apps/api/permissions.py:140]: Not that it's important, just FYI. __str__ could be handled in BaseFlaggedResource as

`return f"`{self.__class__.__name__}` via {self.flag.name}"`

A little less boilerplate and repeated code.

File: apps/api/permissions.py [hunk 2]
```diff
@@ -113,7 +116,31 @@ def is_a_contributor(user):
...
+
+class HasFlaggedAccessToMailChimp(BaseFlaggedResourceAccess):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        Flag = get_waffle_flag_model()
+        self.flag = Flag.objects.filter(name=MAILCHIMP_INTEGRATION_ACCESS_FLAG_NAME).first()
+        if not self.flag:
+            raise ApiConfigurationError()
+
+    def __str__(self):
```
Comment (njharman) [inline apps/api/permissions.py:139]: The __str__ method is uncovered by tests

File: apps/api/permissions.py [hunk 3]
```diff
@@ -113,7 +116,31 @@ def is_a_contributor(user):
...
+        return self.flag.is_active_for_user(request.user) or self.flag.everyone
+
+
+class HasFlaggedAccessToMailChimp(BaseFlaggedResourceAccess):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        Flag = get_waffle_flag_model()
+        self.flag = Flag.objects.filter(name=MAILCHIMP_INTEGRATION_ACCESS_FLAG_NAME).first()
+        if not self.flag:
+            raise ApiConfigurationError()
```
Comment (njharman) [inline apps/api/permissions.py:137]: This raise is uncovered by tests

**Review by njharman — CHANGES_REQUESTED** [review]
Little bit of test coverage missing and covering unrelated line 99 would get this file to 100%! But I don't suppose that possibility gets you as excited as it does I. 

Otherwise looks good.

The AC are written confusing as heck.  I'm assuming (and it looked that way by tests) it is "additive" and not "exclusive" so that if SuperUser is ON, individual non-superuser users with feature flag set ON have access. Despite what AC7 "only superusers can access the Mailchimp connection UI components and APIs." says.

In "simpler to me" words. If any flag gives you access, you have access. If no flag gives you access, you do not have access.  

Right?

File: apps/api/permissions.py (hunk 3 above)
Comment (benjaminEwhite) [reply apps/api/permissions.py:144]: 833d19b4183

File: apps/api/permissions.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/api/permissions.py]: 833d19b4183

File: apps/api/permissions.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/api/permissions.py]: 833d19b4183

**Comment by benjaminEwhite** [comment]
@njharman 

> In "simpler to me" words. If any flag gives you access, you have access. If no flag gives you access, you do not have access.

yes, that's a correct and well put reading of spirit of AC.

**Review by njharman — APPROVED** [review]
Ship it!

File: apps/api/permissions.py [hunk 4]
```diff
@@ -138,3 +130,29 @@ def has_permission(self, request, view):
         (https://github.com/django-waffle/django-waffle/issues/401).
         """
         return self.flag.is_active_for_user(request.user) or self.flag.everyone
+
+    def __str__(self):
+        return f"{self.__class__.__name__} via {self.flag.name if self.flag else '<not configured>'}"
+
+
+class HasFlaggedAccessToMailChimp(BaseFlaggedResourceAccess):
```
Comment (x110dc) [inline apps/api/permissions.py:138]: Terribly minor, but should probably be `HasFlaggedAccessToMailchimp` (lowercase 'c') to match our style guide.

File: apps/api/tests/test_permissions.py [hunk 5]
```diff
@@ -10,3 +18,33 @@ def test_basic(self):
         self._set_up_default_feature_flags()
         t = HasFlaggedAccessToContributionsApiResource()
         assert isinstance(str(t), str)
+
+
+@pytest.mark.django_db
+class TestHasFlaggedAccessToMailChimp:
```
Comment (x110dc) [inline apps/api/tests/test_permissions.py:24]: `chimp`

File: apps/api/tests/test_permissions.py [hunk 6]
```diff
@@ -10,3 +18,33 @@ def test_basic(self):
...
+            pytest_cases.fixture_ref("hub_admin_user"),
+            pytest_cases.fixture_ref("org_user_free_plan"),
+            pytest_cases.fixture_ref("rp_user"),
+        ),
+    )
+    def test_when_flag_set_to_everyone(self, user, default_feature_flags):
+        factory = APIRequestFactory()
+        request = factory.get("/")
+        request.user = user
+        assert HasFlaggedAccessToMailChimp().has_permission(request, None) is True
```
Comment (x110dc) [inline apps/api/tests/test_permissions.py:38]: `chimp` -- I won't flag the rest.

**Review by x110dc — APPROVED** [review]

**Pull request merged**